UnityProject/Assets/Scripts/Items/Devices/AdvancedReleaseGrenade.cs

```diff
@@ -0,0 +1,146 @@
+using System.Collections;
+using Chemistry;
+using Chemistry.Components;
+using UnityEngine;
+using Items.Weapons;
+
+public class AdvancedReleaseGrenade : ChemicalGrenade
+{
+	private int mixAmount = 1;
+
+	public override void MixReagents()
+	{
+		if (isServer)
+		{
+			StartCoroutine(SmartRelease());
+		}
+	}
+
+	IEnumerator SmartRelease()
+	{
+		while (ReagentContainer1.ReagentMixTotal + ReagentContainer2.ReagentMixTotal > 0)
+		{
+			var worldPos = objectPhysics.registerTile.WorldPosition;
+
+			BlastData blastData = new BlastData();
+
+			ReagentMix mixA = ReagentContainer1.CurrentReagentMix.Take(mixAmount);
+			ReagentMix mixB = ReagentContainer1.CurrentReagentMix.Take(mixAmount);
+
+			float internalEnergy = mixA.InternalEnergy + mixB.InternalEnergy;
+
+			ReagentContainer1.ReagentsChanged(true);
+			ReagentContainer1.OnReagentMixChanged?.Invoke();
+			ReagentContainer2.ReagentsChanged(true);
+			ReagentContainer2.OnReagentMixChanged?.Invoke();
+
+			mixedReagentContainer.CurrentReagentMix.InternalEnergy = internalEnergy;
+			mixedReagentContainer.CurrentReagentMix.Temperature += TemperatureChange;
+			mixedReagentContainer.ReagentsChanged(false, true); //We mix the the two containers, but cache the effects of the mixed container.
+
+			blastData.ReagentMix = mixedReagentContainer.CurrentReagentMix.CloneWithCache();
+			ExplosiveBase.ExplosionEvent.Invoke(worldPos, blastData);
+
+			mixedReagentContainer.CurrentReagentMix.ApplyEffectCache(mixedReagentContainer);//We now apply the cache
+			mixedReagentContainer.OnReagentMixChanged?.Invoke();
+			spriteHandler.SetCatalogueIndexSprite(LOCKED_SPRITE);
+			spriteHandler.SetSpriteVariant(EMPTY_VARIANT);
+			mixedReagentContainer.Spill(objectPhysics.OfficialPosition.CutToInt(), DETONATE_SPILL_AMOUNT);
+
+			yield return new WaitForSeconds(1f);
```

Comment:
Needs a fail safe in case the reagent container contains a funny number. Make it so that if it keeps spilling for 30 seconds, it exists out of the while loop.

Reply:
Also use WairFor.Seconds() as it doesn't GC